### PR TITLE
Improve Java and JavaDoc parser by inspecting Maven module

### PR DESCRIPTION
**Pull request by Jagrutiti**
Maven uses the same warnings format for the compiler and the JavaDoc post processing. That means that both the Java and JavaDoc parser will pick up the same warnings. Since warnings in Maven are separated by the actual goal that produces these warnings we simply can improve both parsers by filtering the warnings that are part of the corresponding maven goal.

Wrote a test case to expose the bug. Working on them to fix them.

# Issue

- [JENKINS-67521](https://issues.jenkins.io/browse/JENKINS-67521)
- [JENKINS-63346](https://issues.jenkins.io/browse/JENKINS-63346)

### CODE REVIEW

File: src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java [hunk 1]
```diff
@@ -274,5 +274,11 @@ private void assertThatWarningIsAtLine(final Issue annotation, final int lineNum
                             "/home/hudson/hudson/data/jobs/Hudson main/workspace/remoting/src/test/java/hudson/remoting/BinarySafeStreamTest.java");
         }
     }
+
+    @Test
```
Comment (Jagrutiti) [inline src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java:278]: Hey @uhafner 

When I run this test case. I get the following output:

```
java.lang.AssertionError: 
Expecting size of:
  <- (-): 0 issues (0 duplicates)>
to be:
  <1>
but was:
  <0>

```
There is only one `maven-compiler-plugin` in the `tracker_issue63346.log` file. So it should have only one line. 

Does this mean the bug is exposed? 

Where shall I fix it?

Comment (uhafner) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java:278]: The bug is not yet exposed. As @KalleOlaviNiemitalo already mentioned you need to replace the 2 place holder texts with an actual warning from [JENKINS-67521](https://issues.jenkins.io/browse/JENKINS-67521) (or https://issues.jenkins.io/browse/JENKINS-63346). (Not that you cannot use the same warnings in both blocks since duplicates are removed). Then your test should pick two actual warnings  (and in JavaDocParserTest) now. Expected count is 1. When that is done, you can fix both parsers, so that the expected will be the actual.

File: src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java [hunk 2]
```diff
@@ -277,8 +277,10 @@ private void assertThatWarningIsAtLine(final Issue annotation, final int lineNum
 
     @Test
     void shouldParseJavaWarningsInMavenCompilerPlugin(){
-        Report file = parse("tracker_issue63346.log");
-        assertThat(file).hasSize(1);
+        Report warnings = parse("tracker_issue63346.log");
```
[the remark this answers is not in the document]
Comment (Jagrutiti) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java:280]: When I used `maven.txt` as input file. It recognized the warnings. Is it okay if  I use those warnings?

File: src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java (hunk 1 above)
Comment (Jagrutiti) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java:278]: After this line:

`[INFO] --- maven-compiler-plugin:3.10.1:testCompile (default-testCompile) @ analysis-model ---`

JavacParser warnings will be recognized?

And after this line:

`[INFO] Building jar: /home/jenkins/workspace/Plugins_analysis-model_master/target/analysis-model-10.23.0-rc4948.f52d79122c41-javadoc.jar`

JavacDoc warning should be recognized?

Am I correct in my understanding?

File: src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jagrutiti) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java:280]: Hey @uhafner , @KalleOlaviNiemitalo 

Could you tell me if the bug is exposed now? 

I know you must be expecting me to understand this on my own.

I am confused. I do not know how the actual info, errors, and warnings are displayed by a compiler and in what order. 

It is fascinating to work at this level of the product. At the same time challenging too.

Thank you helping me understand the problem and derive to a solution.

File: src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java [hunk 3]
```diff
@@ -274,5 +274,13 @@ private void assertThatWarningIsAtLine(final Issue annotation, final int lineNum
                             "/home/hudson/hudson/data/jobs/Hudson main/workspace/remoting/src/test/java/hudson/remoting/BinarySafeStreamTest.java");
         }
     }
+
+    @Test
+    void shouldParseJavaWarningsInMavenCompilerPlugin(){
+        Report warnings = parse("tracker_issue63346.log");
+        assertThat(warnings).hasSize(4);
+        assertThat(warnings.get(0)).hasSeverity(Severity.WARNING_NORMAL);
```
[the remark this answers is not in the document]
Comment (Jagrutiti) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: Hey @KalleOlaviNiemitalo 

I have added assert to JavacParser. It asserts all 5 warnings which is the bug. 

But the JavaDocParserTest does not parse a single warning. Ideally it should recognize 2 of them. The assertion fail when I assert `hasSize(2)`. It is expecting size 0 now.

[the remark this answers is not in the document]
Comment (Jagrutiti) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: Where I should fix the JavacParserTest bug? 

I can work on this, right? Because the bug is exposed now?

File: src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java (hunk 1 above)
Comment (uhafner) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: Almost. The compile step can be in different goals, so it might make sense to check for `maven-compiler-plugin` only (without `testCompile`).

File: src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jagrutiti) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: Hey @KalleOlaviNiemitalo 

I have moved the test from the JavacParserTest to AntJavacParserTest. 

As you have mentioned it only parsed lines without parenthesis for line numbers.

The required fix is, to change the `ANT_JAVAC_WARNING_PATTERN`
https://github.com/jenkinsci/analysis-model/blob/b5551d896fc60f0b5886be365695aeff84be4614/src/main/java/edu/hm/hafner/analysis/parser/AntJavacParser.java#L24-L28

to parse warnings that have parenthesis in line number?

And `JavaDocParserTest` should parse warnings without parenthesis in line number? For now, my JavaDocParserTest does not parse any lines.

[the remark this answers is not in the document]
Comment (uhafner) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: Seems that the output of the warnings of javac and the javadoc tool changed in recent versions. Sorry about that, I wasn't aware of those problems.

So I fixed the log file for this test and adapted the number of warnings that I am expecting for each of the three parsers. I did not yet change the assertions after the size check, this is something you can add @Jagrutiti. 

The new log is from a recent compile step here on my machine using Java 17. 
I would expect it to have 4 compiler warnings (JavacParser) and 3 java doc warning (JavaDocParser) and 0 AntJavacParser warnings.

Next steps:
- change all three parsers so that the number of warnings is correct.
- adapt the assertions after the size check.

[the remark this answers is not in the document]
Comment (Jagrutiti) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: Hey @uhafner 

The assertions are not what you expect.

```
[ERROR] Failures: 
[ERROR]   AntJavacParserTest.shouldParseJavaWarningsInMavenCompilerPlugin:231 
Expecting size of:
  <- (-): 2 issues (0 duplicates)>
to be:
  <0>
but was:
  <2>
[ERROR]   JavaDocParserTest.shouldParseJavaDocsWarningsWithMavenSourcePlugin:318 
Expecting size of:
  <- (-): 1 issues (0 duplicates)>
[ERROR]

```

`JavacParser `has 4 warnings as expected.

But, 
`JavaDocParserTest` has 1 warning size.
`AntJavacParserTest` has 2 warnings.

How shall I proceed now?

[the remark this answers is not in the document]
Comment (Jagrutiti) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: Hey @KalleOlaviNiemitalo , @uhafner 

I have updated the test so that they pass. 

Asserted them with fileName for a better understanding of what the test case expects.

>I would expect it to have 4 compiler warnings (JavacParser) and 3 java doc warning (JavaDocParser) and 0 AntJavacParser warnings.

According to this. I need the bug is in JavaDocParser, and AntJavacParser?

File: src/test/java/edu/hm/hafner/analysis/parser/AntJavacParserTest.java [hunk 4]
```diff
@@ -228,7 +228,15 @@ protected AntJavacParser createParser() {
     @Test
     void shouldParseJavaWarningsInMavenCompilerPlugin() {
         Report warnings = parse("issue63346.log");
-        assertThat(warnings).hasSize(0);
+        assertThat(warnings).hasSize(2);
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/AntJavacParserTest.java:231]: Why did you change that back? This parser is broken and should not report a warning.

File: src/test/java/edu/hm/hafner/analysis/parser/JavaDocParserTest.java [hunk 5]
```diff
@@ -315,15 +315,10 @@ void issue55805() {
     @Test
     void shouldParseJavaDocsWarningsWithMavenSourcePlugin() {
         Report warnings = parse("issue63346.log");
-        assertThat(warnings).hasSize(3);
+        assertThat(warnings).hasSize(1);
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/JavaDocParserTest.java:318]: Why did you change that back? This parser is broken and should report 3 warnings:

```
[INFO] --- maven-javadoc-plugin:3.4.1:javadoc (default-cli) @ codingstyle ---
[INFO] Configuration changed, re-generating javadoc.
[WARNING] Javadoc Warnings
[WARNING] /Users/hafner/git/warnings-ng-plugin-devenv/codingstyle/src/main/java/edu/hm/hafner/util/FilteredLog.java:135: warning: no @return
[WARNING] public int size() {
[WARNING] ^
[WARNING] /Users/hafner/git/warnings-ng-plugin-devenv/codingstyle/src/main/java/edu/hm/hafner/util/LookaheadStream.java:49: warning: no comment
[WARNING] public String getFileName() {
[WARNING] ^
[WARNING] /Users/hafner/git/warnings-ng-plugin-devenv/codingstyle/src/main/java/edu/hm/hafner/util/ResourceExtractor.java:70: warning: no comment
[WARNING] public String getResourcePath() {
[WARNING] ^
[WARNING] 3 warnings
```

File: src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (uhafner) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: The reasoning of TDD is to create a test that *fails* and then change the code so that the red test will turn green. So you need to restore my original assertions, fix JavaDocParser, and AntJavacParser and then the tests should all get green. (And the existing test should stay green).

**Comment by uhafner** [comment]
The JavaDoc parser need to get the line 41 changed:

```
 return line.contains("error") && line.contains("ERROR");
```

This should also return true for warnings.

File: src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jagrutiti) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: Hey @uhafner 

I have rectified the two test cases to fail.

I understand the following reasons for the failure:

-  AntJavacParser shouldn't parse the JavaDoc warnings. 
-  JavaDocParser should parse all its 3 warnings. But is parses only one.

I was trying to find the difference between the warnings when  I found that
JavaDocParser test parsed the line with `@` before the last word all the test cases turn green.

The following is the sample input that turns them green:
```
[WARNING] Javadoc Warnings
[WARNING] /Users/hafner/git/warnings-ng-plugin-devenv/codingstyle/src/main/java/edu/hm/hafner/util/FilteredLog.java:135: warning: no @return
[WARNING] public int size() {
[WARNING] ^
[WARNING] /Users/hafner/git/warnings-ng-plugin-devenv/codingstyle/src/main/java/edu/hm/hafner/util/LookaheadStream.java:49: warning: no @comment
[WARNING] public String getFileName() {
[WARNING] ^
[WARNING] /Users/hafner/git/warnings-ng-plugin-devenv/codingstyle/src/main/java/edu/hm/hafner/util/ResourceExtractor.java:70: warning: no @comment
[WARNING] public String getResourcePath() {
[WARNING] ^
[WARNING] 3 warnings
```

While in the current input the lines are without `@`:

```

[WARNING] /Users/hafner/git/warnings-ng-plugin-devenv/codingstyle/src/main/java/edu/hm/hafner/util/LookaheadStream.java:49: warning: no comment

[WARNING] /Users/hafner/git/warnings-ng-plugin-devenv/codingstyle/src/main/java/edu/hm/hafner/util/ResourceExtractor.java:70: warning: no comment

```

Does that means
JavaDocParser regex should look for `@`?

And AntJavacParser,
regex should ignore `@`?

or should I write a method for these fixes?

What is the ideal solution here? 
Is my speculation correct?

[the remark this answers is not in the document]
Comment (uhafner) [reply src/test/java/edu/hm/hafner/analysis/parser/JavacParserTest.java]: The best way would be to create a new subclass `AbstractMavenLogParser` for `MavenConsoleParser`, `AntJavacParser`, `JavaDocParser`, `JavacParser`: this subclass should use the code in `MavenConsoleParser` to detect the plugin and goal. 

Then:  
- `AntJavacParser` should ignore all warnings, if a maven goal is set (in `isLineInteresting`). Otherwise proceed as normal. 
- `JavaDocParser`, `JavacParser` should ignore warnings, that are from a different maven plugin.
- `MavenConsoleParser` should work unchanged, just the  goal and plugin detection will be from the parent class

File: src/main/java/edu/hm/hafner/AbstractMavenLogParser.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * A parser subclass for MavenConsoleParser, AntJavacParser, JavaDocParser, JavacParser
+ *
+ * @author Jagruti Tiwari
+ */
+
+public class AbstractMavenLogParser extends MavenConsoleParser {
+
+    @Override
+    protected boolean isLineInteresting(final String line) {
```
Comment (Jagrutiti) [inline src/main/java/edu/hm/hafner/AbstractMavenLogParser.java:14]: Hey @uhafner ,

For each of the parsers where should I write the code to check the following:

> 
>     * `AntJavacParser` should ignore all warnings, if a maven goal is set (in `isLineInteresting`). Otherwise proceed as normal.
> 
>     * `JavaDocParser`, `JavacParser` should ignore warnings, that are from a different maven plugin.
> 
>     * `MavenConsoleParser` should work unchanged, just the  goal and plugin detection will be from the parent class

**Review by Jagrutiti — COMMENTED** [review]
Hey @uhafner,

This change worked for fixing `JavaDocParserTest`. 

I do not know if this is the fix you had in mind.

> The best way would be to create a new subclass `AbstractMavenLogParser` for `MavenConsoleParser`, `AntJavacParser`, `JavaDocParser`, `JavacParser`: this subclass should use the code in `MavenConsoleParser` to detect the plugin and goal.

Still I am unable to figure out how to use the subclass to detect the goal and fix `AntJavacParserTest`.

**Comment by uhafner** [comment]
> Hey @uhafner,
> 
> This change worked for fixing `JavaDocParserTest`.
> 
> I do not know if this is the fix you had in mind.
> 
> > The best way would be to create a new subclass `AbstractMavenLogParser` for `MavenConsoleParser`, `AntJavacParser`, `JavaDocParser`, `JavacParser`: this subclass should use the code in `MavenConsoleParser` to detect the plugin and goal.
> 
> Still I am unable to figure out how to use the subclass to detect the goal and fix `AntJavacParserTest`.

The Ant parser should skip all warnings that are part of a Maven build. So you first need to detect the Maven steps. As mentioned above, this code is already part of `MavenConsoleParser`. You need to move that code to a new base class. So that all other classes can use it.

File: src/main/java/edu/hm/hafner/analysis/parser/AbstractMavenLogParser.java [hunk 7]
```diff
@@ -12,17 +15,52 @@
...
-        return super.isLineInteresting(line);
+    @Override
+    protected boolean isLineInteresting(final String line) {
+        Matcher goalMatcher = MAVEN_PLUGIN_START.matcher(line);
+
+        if (goalMatcher.find()) {
+            goal = String.format("%s:%s", goalMatcher.group("id"), goalMatcher.group("goal"));
+        }
+
+        if(!goal.isEmpty()){
```
Comment (Jagrutiti) [inline src/main/java/edu/hm/hafner/analysis/parser/AbstractMavenLogParser.java:45]: Hey @uhafner 


Could you confirm if I have shifted the correct code  from `MavnConsoleParser`.

Because though the `AntJavacParserTest` succeeds, other 10 test casess fail :(

**Comment by uhafner** [comment]
Your inheritance is not correct. The new Abstract parser should be top level. This class should just locate and store the maven IDs. All other parsers should derive from this new class!

Abstract <|- Ant
Abstract <|- Maven
Abstract <|- JavaC
Abstract <|- JavaDoc

**Review by Jagrutiti — COMMENTED** [review]
Hey @uhafner 

I noticed all the derived classes have extended the `LookaheadParser` class.

>Abstract <|- Ant
>Abstract <|- Maven
>Abstract <|- JavaC
>Abstract <|- JavaDoc

So, in my current code, `AbstractMavenLogParser extends LookaheadParser`.

And all the other classes extend `AbstractMavenLogParser`.

But the problems do not end. 

On running `mvn clean verify` this is my issue list:

`[ERROR] Tests run: 1075, Failures: 16, Errors: 2, Skipped: 1`

Failures have increased from 10 to 16.

Now, I know I have not shifted a lot of code from derived classes. 

What should be my criteria to decide what code to shift?

**Comment by uhafner** [comment]
I fixed the base class so that the tests succeed now. The solution was a little bit more complex than I thought in the beginning, sorry about that. You should have started with the first issue that I proposed and not the last one 😉

**Comment by Jagrutiti** [comment]
> I fixed the base class so that the tests succeed now. The solution was a little bit more complex than I thought in the beginning, sorry about that. You should have started with the first issue that I proposed and not the last one 😉

Thank you for fixing the base class for me.

Does this mean this issue is ready to close?

Do you want me to do anything else here?

**Comment by uhafner** [comment]
> > I fixed the base class so that the tests succeed now. The solution was a little bit more complex than I thought in the beginning, sorry about that. You should have started with the first issue that I proposed and not the last one 😉
> 
> Thank you for fixing the base class for me.
> Does this mean this issue is ready to close?

Not yet: in order to merge the PR all checks need to be green. There are still some open PMD and SpotBugs issues.

> 
> Do you want me to do anything else here?

One thing that would be helpful (even though a little bit out of scope of this PR topic): the JavaDoc parser should report warnings of type `ERROR` to the severity `Error`, all others to severity Warning_Normal. Currently errors are mapped to HIGH.

**Comment by Jagrutiti** [comment]
> Not yet: in order to merge the PR all checks need to be green. There are still some open PMD and SpotBugs issues.

I has used Spotbugs when [modernizing one of the plugins
](https://www.jenkins.io/doc/developer/tutorial-improve/add-more-spotbugs-checks/) but now this command `$ mvn clean -DskipTests verify` passes. 

Should I try to resolve warnings like these: 
```
  Did you mean 'var spotBugs = new Report(SPOTBUGS_ID, SPOTBUGS_NAME, "spotbugs.xml");'?
[WARNING] C:\Users\tiwar\open-source\analysis-model\src\test\java\edu\hm\hafner\analysis\ReportTest.java:[910,15] [Varifier] Consider using `var` here to avoid boilerplate.
    (see https://errorprone.info/bugpattern/Varifier)
  Did you mean 'var wrappedCheckStyle = new Report();'?

```

How do I find those errors?

> One thing that would be helpful (even though a little bit out of scope of this PR topic): the JavaDoc parser should report warnings of type `ERROR` to the severity `Error`, all others to severity Warning_Normal. Currently errors are mapped to HIGH.

You want me to change[ lines like this](https://github.com/Jagrutiti/analysis-model/blob/dc2c73a3ba4593adda6ce2bdb1995ee563ec630e/src/main/java/edu/hm/hafner/analysis/parser/PuppetLintParser.java#L50) to `Severity.ERROR` ?
https://github.com/Jagrutiti/plugin-installation-manager-tool/blob/b7245ace1620d3b718ae5374d7440b6280d7f9e3/plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java#L1498
The line checking ` (level.contains("error") || level.contains("ERROR")`

**Comment by uhafner** [comment]
> > Not yet: in order to merge the PR all checks need to be green. There are still some open PMD and SpotBugs issues.
> 
> I has used Spotbugs when [modernizing one of the plugins ](https://www.jenkins.io/doc/developer/tutorial-improve/add-more-spotbugs-checks/) but now this command `$ mvn clean -DskipTests verify` passes.
> 
> Should I try to resolve warnings like these:
> 
> ```
>   Did you mean 'var spotBugs = new Report(SPOTBUGS_ID, SPOTBUGS_NAME, "spotbugs.xml");'?
> [WARNING] C:\Users\tiwar\open-source\analysis-model\src\test\java\edu\hm\hafner\analysis\ReportTest.java:[910,15] [Varifier] Consider using `var` here to avoid boilerplate.
>     (see https://errorprone.info/bugpattern/Varifier)
>   Did you mean 'var wrappedCheckStyle = new Report();'?
> ```
> 
> How do I find those errors?

Simply look into the pull request checks details right below this text. Here is a screenshot:

Bildschirm­foto 2023-01-29 um 19 24 53


> 
> > One thing that would be helpful (even though a little bit out of scope of this PR topic): the JavaDoc parser should report warnings of type `ERROR` to the severity `Error`, all others to severity Warning_Normal. Currently errors are mapped to HIGH.
> 
> You want me to change[ lines like this](https://github.com/Jagrutiti/analysis-model/blob/dc2c73a3ba4593adda6ce2bdb1995ee563ec630e/src/main/java/edu/hm/hafner/analysis/parser/JavaDocParser.java#L81) to `Severity.ERROR` in `JavaDocParser` ?

Exactly this line, no other line. (Hopefully a test fails after the change).
> 
> The lines checking ` (level.contains("error") || level.contains("ERROR")`.
> 

This is correct. 

> Is there some easy way to find such lines? Other than whole project search?

No, just in the JavaDoc parser.
> 
> I found only one such change in JavaDocParser? Are there other files where I have to change?

No

File: src/test/java/edu/hm/hafner/analysis/parser/JavaDocParserTest.java [hunk 8]
```diff
@@ -78,23 +78,23 @@ void issue37975() {
 
         try (SoftAssertions softly = new SoftAssertions()) {
             softly.assertThat(warnings.get(0))
-                    .hasSeverity(Severity.WARNING_HIGH)
+                    .hasSeverity(Severity.ERROR)
```
Comment (Jagrutiti) [inline src/test/java/edu/hm/hafner/analysis/parser/JavaDocParserTest.java:81]: Two test cases failed when I change to `Severtity.ERROR`.

**Review by uhafner — APPROVED** [review]
Thanks! Now everything is fine.
